apps/namadillo/src/atoms/balance/atoms.ts

```diff
@@ -0,0 +1,230 @@
+  const defaultAccountQuery = get(defaultAccountAtom);
+
+  return {
+    queryKey: ["viewing-key", accountsQuery.data, defaultAccountQuery.data],
+    ...queryDependentFn(async () => {
+      const shieldedAccount = accountsQuery.data?.find(
+        (a) => a.isShielded && a.alias === defaultAccountQuery.data?.alias
+      );
+      return shieldedAccount?.viewingKey ?? "";
+    }, [accountsQuery, defaultAccountQuery]),
+  };
+});
+
+export const shieldedBalanceAtom = atomWithQuery<Balance>((get) => {
+  const enablePolling = get(shouldUpdateBalanceAtom);
+  const viewingKeyQuery = get(viewingKeyAtom);
+  const tokenAddressesQuery = get(tokenAddressesAtom);
+
+  return {
+    refetchInterval: enablePolling ? 1000 : false,
+    queryKey: [
+      "shielded-balance",
+      viewingKeyQuery.data,
+      tokenAddressesQuery.data,
+    ],
+    ...queryDependentFn(async () => {
+      const viewingKey = viewingKeyQuery.data;
+      const tokenAddresses = tokenAddressesQuery.data;
+      if (!viewingKey || !tokenAddresses) {
+        return [];
+      }
+      // TODO mock
+      // await new Promise((r) => setTimeout(() => r(0), 500));
+      // getSdkInstance().then((sdk) => sdk.rpc.shieldedSync([viewingKey]));
+      // return [
+      //   ["tnam1qy440ynh9fwrx8aewjvvmu38zxqgukgc259fzp6h", "37"], // nam
+      //   ["tnam1p5nnjnasjtfwen2kzg78fumwfs0eycqpecuc2jwz", "1"], // uatom
+      // ];
+      const sdk = await getSdkInstance();
+      await sdk.rpc.shieldedSync([viewingKey]);
+      return await sdk.rpc.queryBalance(
+        viewingKey,
+        tokenAddresses.map((t) => t.address)
+      );
+    }, [viewingKeyQuery, tokenAddressesQuery]),
+  };
+});
+
+export const denomByAddressAtom = atomWithQuery((get) => {
+  const tokenAddressesQuery = get(tokenAddressesAtom);
+  const namTokenAddressQuery = get(nativeTokenAddressAtom);
+
+  return {
+    queryKey: [
+      "denom-by-address",
+      tokenAddressesQuery.data,
+      namTokenAddressQuery.data,
+    ],
+    queryFn: () => {
+      const addressToDenom: Record<string, string> = {};
+      if (namTokenAddressQuery.data) {
+        addressToDenom[namTokenAddressQuery.data] = NAM_DENOM;
+      }
+      tokenAddressesQuery.data?.forEach((token) => {
+        if ("trace" in token) {
+          const denom = token.trace.split("/").at(-1) ?? token.trace;
+          addressToDenom[token.address] = denom;
+        }
+      });
+      return addressToDenom;
+    },
+  };
+});
+
+export const assetsByDenomAtom = atom((get) => {
+  const availableAssets = get(availableAssetsAtom);
+  const assetsByDenom: Record<string, Asset> = {
+    // TODO namAsset should be returned from availableAssetsAtom
+    [NAM_DENOM]: namadaAsset,
+  };
+  availableAssets.forEach((assets) => {
+    assets.assets.forEach((asset) => {
+      asset.denom_units.forEach((unit) => {
+        // only set the asset if it doesn't exists
+        // otherwise the testnet will override the mainnet
```

Comment:
I just realised both this PR and #1228 are trying to decode assets from their IBC trace, but in different ways. Maybe we can use just one way, but we'll need to decide which way.

Like you said in this comment, if the asset exists on multiple chains, we will only use the first entry, but the problem with this is that different chains have different configurations for the same asset (and not just testnets).

I'll write it up a bit more in a message so we can discuss it.